tests/Go/TestUtils/JoinPointsExtractor.php

```diff
@@ -0,0 +1,116 @@
+<?php
+/*
+ * Go! AOP framework
+ *
+ * @copyright Copyright 2011, Lisachenko Alexander <lisachenko.it@gmail.com>
+ *
+ * This source file is subject to the license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Go\TestUtils;
+
+use PhpParser\Node;
+use PhpParser\NodeTraverser;
+use PhpParser\Node\Stmt;
+use PhpParser\NodeVisitorAbstract;
+use PhpParser\Parser;
+use PhpParser\ParserFactory;
+use PhpParser\PrettyPrinter\Standard as Printer;
+
+/**
+ * Extracts join points definitions from woven class proxy by parsing and traversing class AST.
+ */
+final class JoinPointsExtractor extends NodeVisitorAbstract
+{
+    /**
+     * @var null|Node\Expr\Array_
+     */
+    private $joinPoints = null;
+
+    /**
+     * @var null|Parser
+     */
+    private static $parser;
+
+    /**
+     * @var null|NodeTraverser
+     */
+    private static $traverser;
+
+    /**
+     * @var null|JoinPointsExtractor
+     */
+    private static $extractor;
+
+    /**
+     * @var null|Printer
+     */
+    private static $printer;
+
+    /**
+     * {@inheritdoc}
+     */
+    public function beforeTraverse(array $nodes)
+    {
+        $this->joinPoints = null;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function enterNode(Node $node)
+    {
+        if (
+            $node instanceof Stmt\Property
+            &&
+            $node->isPrivate()
+            &&
+            $node->isStatic()
+            &&
+            '__joinPoints' === $node->props[0]->name
+        ) {
+            $this->joinPoints = $node->props[0]->default;
+        }
+    }
+
+    /**
+     * Get extracted value of private static "__joinPoints" property from woven class.
+     *
+     * @return Node\Expr\Array_
+     */
+    public function getExtractedAst()
+    {
+        return $this->joinPoints;
+    }
+
+    /**
+     * Extract join points from woven class
+     *
+     * @param string $class Either PHP code of class, or path to class file.
+     *
+     * @return array Value of private static "__joinPoints" property.
+     */
+    public static function extractJoinPoints($class)
+    {
+        if (file_exists($class)) {
+            $class = file_get_contents($class);
+        }
+
+        if (null === self::$parser) {
+            self::$parser    = (new ParserFactory())->create(ParserFactory::PREFER_PHP7);
+            self::$traverser = new NodeTraverser();
+            self::$extractor = new JoinPointsExtractor();
+            self::$printer   = new Printer();
+
+            self::$traverser->addVisitor(self::$extractor);
+        }
+
+        $stmts = self::$parser->parse($class);
+        self::$traverser->traverse($stmts);
+
+        $ast = self::$extractor->getExtractedAst();
+
+        return eval('return ' . self::$printer->prettyPrint([$ast]));
```

Comment:
Eval is evil if it used without caution. 

Here -> `eval` is used having in mind its intent and purpose and it is used properly. It is safe because source of code for `eval` is known and predictable. On top of that I have to be honest, I do not know how to evaluate PHP code and load it into memory (after I have extracted desired part of AST) without using `eval`.  

If you know different method - no problemo, will implement that!

Now, you have to first make decision wether you are going to allow extraction of join points without loading woven class into memory. After that decision, we can discuss about using `eval` here. 

BTW: https://github.com/sebastianbergmann/phpunit-mock-objects/blob/master/src/Generator.php#L263